flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,282 @@
+
+        val weightedSum = poolSubArray.map(lp => lp._1 * lp._3).sum
+        val weight = poolSubArray.map(_._3).sum
+
+        var i = start
+        while (i <= end) {
+          input(i) = (weightedSum / weight, input(i)._2, input(i)._3)
+          i = i + 1
+        }
+      }
+
+      var i = 0
+      val len = input.length
+      while (i < len) {
+        var j = i
+
+        // Find monotonicity violating sequence, if any.
+        while (j < len - 1 && input(j)._1 > input(j + 1)._1) {
+          j = j + 1
+        }
+
+        // If monotonicity was not violated, move to next data point.
+        if (i == j) {
+          i = i + 1
+        } else {
+          // Otherwise pool the violating sequence
+          // and check if pooling caused monotonicity violation in previously processed
+          // points.
+          while (i >= 0 && input(i)._1 > input(i + 1)._1) {
+            pool(input, i, j)
+            i = i - 1
+          }
+
+          i = j
+        }
+      }
+      // For points having the same prediction, we only keep two boundary points.
+      val compressed = ArrayBuffer.empty[(Double, Double, Double)]
+
+      var (curLabel, curFeature, curWeight) = input.head
+      var rightBound = curFeature
+      def merge(): Unit = {
+        compressed += ((curLabel, curFeature, curWeight))
+        if (rightBound > curFeature) {
+          compressed += ((curLabel, rightBound, 0.0))
+        }
+      }
+      i = 1
+      while (i < input.length) {
+        val (label, feature, weight) = input(i)
+        if (label == curLabel) {
+          curWeight += weight
+          rightBound = feature
+        } else {
+          merge()
+          curLabel = label
+          curFeature = feature
+          curWeight = weight
+          rightBound = curFeature
+        }
+        i += 1
+      }
+      merge()
+
+      compressed.toArray
+    }
+  }
+
+
+  implicit val fitIR = new FitOperation[IsotonicRegression, (Double, Double, Double)] {
+
+    override def fit(instance: IsotonicRegression,
+                     fitParameters: ParameterMap,
+                     input: DataSet[(Double, Double, Double)]): Unit = {
+
+      val isotonic = instance.isotonic
+
+      val preprocessedInput = if (isotonic) {
+        input
+      } else {
+        input.map(x => (-x._1, x._2, x._3))
+      }
+
+      val parallelStepResult = preprocessedInput
+        .partitionByRange(1)
+        .mapPartition(partition => {
+          val buffer = new ArrayBuffer[(Double, Double, Double)]
+          buffer ++= partition
+          Seq(buffer.sortBy(x => (x._2, x._1)).toArray)
+        })
+        .map(new AdjacentPoolViolatersMapper)
+        .setParallelism(1)
+        .mapPartition(partitions => {
+          val buffer = new ArrayBuffer[(Double, Double, Double)]
+          for (partition <- partitions) {
+            buffer ++= partition
+          }
+          Seq(buffer.sortBy(x => (x._2, x._1)).toArray)
```

Comment:
This operation effectively collects all data on a single machine and materializes it. This won't scale at all.